// pw2/lib/cjs/index.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.isPOW2Value = exports.getPreviousPOW2Value = exports.getNextPOW2Value = exports.getNearestPOW2Value = exports.createPw2 = exports.RESIZING_MODES = void 0;
const resizingModes_js_1 = __importDefault(require("./resizingModes.js"));
exports.RESIZING_MODES = resizingModes_js_1.default;
const pw2_1 = __importDefault(require("./pw2"));
exports.createPw2 = pw2_1.default;
const PW2ValueConverter_js_1 = require("./PW2ValueConverter.js");
Object.defineProperty(exports, "getNearestPOW2Value", { enumerable: true, get: function () { return PW2ValueConverter_js_1.getNearestPOW2Value; } });
Object.defineProperty(exports, "getNextPOW2Value", { enumerable: true, get: function () { return PW2ValueConverter_js_1.getNextPOW2Value; } });
Object.defineProperty(exports, "getPreviousPOW2Value", { enumerable: true, get: function () { return PW2ValueConverter_js_1.getPreviousPOW2Value; } });
Object.defineProperty(exports, "isPOW2Value", { enumerable: true, get: function () { return PW2ValueConverter_js_1.isPOW2Value; } });
